src/servers/src/proto.rs

```diff
@@ -0,0 +1,295 @@
+use crate::repeated_field::{Clear, RepeatedField};
+
+impl Clear for Sample {
+    fn clear(&mut self) {}
+}
+
+#[derive(Default, Clone)]
+pub struct PromLabel {
+    pub name: Bytes,
+    pub value: Bytes,
+}
+
+impl Clear for PromLabel {
+    fn clear(&mut self) {
+        self.name.clear();
+        self.value.clear();
+    }
+}
+
+impl PromLabel {
+    pub fn merge_field<B>(
+        &mut self,
+        tag: u32,
+        wire_type: WireType,
+        buf: &mut B,
+        ctx: DecodeContext,
+    ) -> Result<(), DecodeError>
+    where
+        B: Buf,
+    {
+        const STRUCT_NAME: &str = "PromLabel";
+        match tag {
+            1u32 => {
+                let value = &mut self.name;
+                prost::encoding::bytes::merge(wire_type, value, buf, ctx).map_err(|mut error| {
+                    error.push(STRUCT_NAME, "name");
+                    error
+                })
+            }
+            2u32 => {
+                let value = &mut self.value;
+                prost::encoding::bytes::merge(wire_type, value, buf, ctx).map_err(|mut error| {
+                    error.push(STRUCT_NAME, "value");
+                    error
+                })
+            }
+            _ => prost::encoding::skip_field(wire_type, tag, buf, ctx),
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct PromTimeSeries {
+    pub table_name: String,
+    pub labels: RepeatedField<PromLabel>,
+    pub samples: RepeatedField<Sample>,
+}
+
+impl Clear for PromTimeSeries {
+    fn clear(&mut self) {
+        self.table_name.clear();
+        self.labels.clear();
+        self.samples.clear();
+    }
+}
+
+impl PromTimeSeries {
+    pub fn merge_field<B>(
+        &mut self,
+        tag: u32,
+        wire_type: WireType,
+        buf: &mut B,
+        ctx: DecodeContext,
+    ) -> Result<(), DecodeError>
+    where
+        B: Buf,
+    {
+        const STRUCT_NAME: &str = "PromTimeSeries";
+        match tag {
+            1u32 => {
+                // decode labels
+                let label = self.labels.push_default();
+
+                let len = decode_varint(buf).map_err(|mut error| {
+                    error.push(STRUCT_NAME, "labels");
+                    error
+                })?;
+                let remaining = buf.remaining();
+                if len > remaining as u64 {
+                    return Err(DecodeError::new("buffer underflow"));
+                }
+
+                let limit = remaining - len as usize;
+                while buf.remaining() > limit {
+                    let (tag, wire_type) = decode_key(buf)?;
+                    label.merge_field(tag, wire_type, buf, ctx.clone())?;
+                }
+                if buf.remaining() != limit {
+                    return Err(DecodeError::new("delimited length exceeded"));
+                }
+                if label.name.deref() == METRIC_NAME_LABEL_BYTES {
+                    let table_name = unsafe { String::from_utf8_unchecked(label.value.to_vec()) };
```

Comment:
nitpick: Maybe we should document unsafe codes.

Reply:
fixed https://github.com/GreptimeTeam/greptimedb/pull/3425/commits/571926fd0e2cda3f14633107fa722bf6e73c55c8#diff-394a4685dfb8084715b4fc86fcbf994dab3d976fb1d4fa8eb9cf51f651548d0eR127